Fix optim state_dict casting to allow step to cast for fused/capturable

janeyx99 opened the pull request:
I'm guessing this should fix https://github.com/pytorch/pytorch/pull/88015#issuecomment-1569523106 but am waiting on @ychfan to supply more details so I could write a good test case.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102619

janeyx99 committed:
Fix optim state_dict casting to allow step to cast to CPU

[ghstack-poisoned]

janeyx99 commented:
Also cc @mikaylagawarecki

janeyx99 committed:
Update on "Fix optim state_dict casting to allow step to cast to CPU"

I'm guessing this should fix https://github.com/pytorch/pytorch/pull/88015#issuecomment-1569523106 but am waiting on ychfan to supply more details so I could write a good test case.




[ghstack-poisoned]

janeyx99 committed:
Update on "Fix optim state_dict casting to allow step to cast to CPU"

I'm guessing this should fix https://github.com/pytorch/pytorch/pull/88015#issuecomment-1569523106 but am waiting on ychfan to supply more details so I could write a good test case.




[ghstack-poisoned]

janeyx99 commented:
Test passes: 
```
2023-06-08T20:25:18.2677106Z test_optim.py::TestOptim::test_fused_optimizer_load_state_dict <- test/optim/test_optim.py PASSED [0.0087s] [  7%]
```

janeyx99 committed:
Update on "Fix optim state_dict casting to allow step to cast to CPU"

I'm guessing this should fix https://github.com/pytorch/pytorch/pull/88015#issuecomment-1569523106 but am waiting on ychfan to supply more details so I could write a good test case.




[ghstack-poisoned]

janeyx99 committed:
Update on "Fix optim state_dict casting to allow step to cast to CPU"

I'm guessing this should fix https://github.com/pytorch/pytorch/pull/88015#issuecomment-1569523106 but am waiting on ychfan to supply more details so I could write a good test case.




[ghstack-poisoned]

albanD commented on torch/optim/optimizer.py:
```diff
@@ -393,6 +394,27 @@ def pack_group(group):
+                break
+
+        if key != "step" or capturable or fused:
+            if param.is_floating_point():
+                value = value.to(param.dtype)
+            value = value.to(param.device)
```
nit but could be significant runtime
```suggestion
            if param.is_floating_point():
                value = value.to(dtype=param.dtype, device=param.device)
            else:
                value = value.to(device=param.device)
```

janeyx99 replied:
oooh good call

albanD reviewed (approved):
Sounds good, only a small suggestion

mikaylagawarecki commented on torch/optim/optimizer.py:
```diff
@@ -393,6 +394,27 @@ def pack_group(group):
         }
 
+    @staticmethod
+    def _process_value_according_to_param_policy(param: Tensor, value: Tensor, paramID: int = None,
```
tiny nit: might be more pythonic to name as `param_id` rather than `paramID` and similarly for other variables ([see here](https://peps.python.org/pep-0008/#function-and-variable-names))

janeyx99 committed:
Update on "Fix optim state_dict casting to allow step to cast to CPU"

I'm guessing this should fix https://github.com/pytorch/pytorch/pull/88015#issuecomment-1569523106 but am waiting on ychfan to supply more details so I could write a good test case.




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

ideasbyjin commented:
Oh it'd be great to have this PR in! FYI, for my particular issue (#104767), I've found using DDP (not DP) had resolved it from HF's end but making this change should be beneficial, I think?

janeyx99 commented:
This PR is already landed (in source and not in the latest release), so the change would need to be picked up by huggingface.

Sakurakdx commented:
I also encountered this problem when I used deepspeed+huggingface trainer to resume. How to solve it? My torch version is 2.0.1
